File: python/advanced_ds/suffix_array.py
"""
Suffix Array with LCP (Longest Common Prefix) Array

Time Complexity:
- Construction: O(n log n) with O(log n) space for sorting
- Pattern Search: O(m log n) where m = pattern length
- LCP Array: O(n) via Kasai's algorithm

Space Complexity: O(n)

Use Cases:
- Substring search (more space-efficient than suffix tree)
- Finding all occurrences of a pattern
- Finding longest repeated substring
- Computing longest common substring
- Approximate string matching

Key Insight:
- Sort all suffixes of the string
- Binary search to find pattern ranges
- LCP array helps in analysis: LCP[i] = longest common prefix of SA[i] and SA[i+1]
- Kasai's algorithm computes LCP in O(n) time
- More cache-friendly than suffix tree
"""

from typing import List, Tuple


class SuffixArray:
    """Suffix array with LCP array using Kasai's algorithm."""

    def __init__(self, text: str):
        """
        Build suffix array and LCP array.

        Args:
            text: Input string
        """
        self.text = text
        self.n = len(text)
        self.sa = []  # Suffix array
        self.lcp = []  # LCP array
        self.rank = []  # Rank array for LCP computation

        if self.n == 0:
            return

        self._build_suffix_array()
        self._build_lcp_array()

    def _build_suffix_array(self) -> None:
        """Build suffix array using simple O(n log² n) approach."""
        # Create pairs of (suffix, original_index)
        suffixes = []
        for i in range(self.n):
            suffixes.append((self.text[i:], i))

        # Sort suffixes
        suffixes.sort()

        # Extract sorted indices
        self.sa = [idx for _, idx in suffixes]

    def _build_lcp_array(self) -> None:
        """Build LCP array using Kasai's algorithm in O(n)."""
        self.rank = [0] * self.n
        for i, sa_idx in enumerate(self.sa):
            self.rank[sa_idx] = i

        self.lcp = [0] * (self.n - 1)
        h = 0

        for i in range(self.n):
            if self.rank[i] > 0:
                j = self.sa[self.rank[i] - 1]
                while i + h < self.n and j + h < self.n and \
                      self.text[i + h] == self.text[j + h]:
                    h += 1

                self.lcp[self.rank[i] - 1] = h

                if h > 0:
                    h -= 1

    def find_pattern(self, pattern: str) -> List[int]:
        """
        Find all occurrences of pattern in text.

        Returns:
            List of starting positions
        """
        if pattern == "" or self.n == 0:
            return []

        # Binary search for pattern in suffix array
        left, right = 0, self.n - 1
        matches = []

        # Find leftmost position where pattern could start
        while left <= right:
            mid = (left + right) // 2
            suffix = self.text[self.sa[mid]:]

            if suffix.startswith(pattern):
                matches.append(self.sa[mid])
                right = mid - 1
            elif pattern < suffix[:len(pattern)]:
                right = mid - 1
            else:
                left = mid + 1

        # Find all matching suffixes
        if matches:
            left = self.sa.index(matches[0])
            for i in range(left + 1, self.n):
                suffix = self.text[self.sa[i]:]
                if suffix.startswith(pattern):
                    matches.append(self.sa[i])
                else:
                    break

        return sorted(matches)

    def longest_repeated_substring(self) -> str:
        """Find longest substring that appears at least twice."""
        if self.n == 0:
            return ""

        max_lcp = 0
        max_idx = -1

        for i, lcp_val in enumerate(self.lcp):
            if lcp_val > max_lcp:
                max_lcp = lcp_val
                max_idx = i

        if max_idx == -1:
            return ""

        start = self.sa[max_idx]
        return self.text[start:start + max_lcp]

    def get_suffix_array(self) -> List[int]:
        """Get the suffix array."""
        return self.sa.copy()

    def get_lcp_array(self) -> List[int]:
        """Get the LCP array."""
        return self.lcp.copy()


if __name__ == "__main__":
    # Example 1: Basic suffix array
    print("=== Example 1: Suffix Array ===")
    text = "banana"
    sa = SuffixArray(text)

    print(f"Text: '{text}'")
    print(f"Suffix Array: {sa.get_suffix_array()}")
    print(f"LCP Array: {sa.get_lcp_array()}")

    print("\nSuffixes in sorted order:")
    for idx in sa.get_suffix_array():
        print(f"  {idx:2d}: {text[idx:]}")

    # Example 2: Pattern finding
    print("\n=== Example 2: Pattern Finding ===")
    text2 = "mississippi"
    sa2 = SuffixArray(text2)

    patterns = ["is", "si", "pp", "mis"]
    print(f"Text: '{text2}'")
    for pattern in patterns:
        positions = sa2.find_pattern(pattern)
        print(f"Pattern '{pattern}': positions {positions}")

    # Example 3: Longest repeated substring
    print("\n=== Example 3: Longest Repeated Substring ===")
    text3 = "abracadabra"
    sa3 = SuffixArray(text3)

    lrs = sa3.longest_repeated_substring()
    print(f"Text: '{text3}'")
    print(f"Longest repeated substring: '{lrs}'")

    text4 = "aabaabaab"
    sa4 = SuffixArray(text4)
    lrs2 = sa4.longest_repeated_substring()
    print(f"\nText: '{text4}'")
    print(f"Longest repeated substring: '{lrs2}'")

File: java/advanced_ds/SuffixArray.java
package advanced_ds;

import java.util.*;

/**
 * Suffix Array with LCP (Longest Common Prefix) Array
 *
 * Time Complexity:
 * - Construction: O(n log n) with O(log n) space for sorting
 * - Pattern Search: O(m log n) where m = pattern length
 * - LCP Array: O(n) via Kasai's algorithm
 *
 * Space Complexity: O(n)
 *
 * Use Cases:
 * - Substring search (more space-efficient than suffix tree)
 * - Finding all occurrences of a pattern
 * - Finding longest repeated substring
 * - Computing longest common substring
 * - Approximate string matching
 *
 * Key Insight:
 * - Sort all suffixes of the string
 * - Binary search to find pattern ranges
 * - LCP array helps in analysis: LCP[i] = longest common prefix of SA[i] and SA[i+1]
 * - Kasai's algorithm computes LCP in O(n) time
 * - More cache-friendly than suffix tree
 */
public class SuffixArray {

    private String text;
    private int n;
    private int[] sa;  // Suffix array
    private int[] lcp; // LCP array
    private int[] rank; // Rank array for LCP computation

    /**
     * Build suffix array and LCP array.
     *
     * @param text Input string
     */
    public SuffixArray(String text) {
        this.text = text;
        this.n = text.length();

        if (n > 0) {
            buildSuffixArray();
            buildLcpArray();
        }
    }

    /**
     * Build suffix array using O(n log² n) approach.
     */
    private void buildSuffixArray() {
        // Create array of suffix indices
        Integer[] indices = new Integer[n];
        for (int i = 0; i < n; i++) {
            indices[i] = i;
        }

        // Sort indices by their suffixes
        Arrays.sort(indices, (i, j) -> {
            return text.substring(i).compareTo(text.substring(j));
        });

        sa = new int[n];
        for (int i = 0; i < n; i++) {
            sa[i] = indices[i];
        }
    }

    /**
     * Build LCP array using Kasai's algorithm in O(n).
     */
    private void buildLcpArray() {
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            rank[sa[i]] = i;
        }

        lcp = new int[n - 1];
        int h = 0;

        for (int i = 0; i < n; i++) {
            if (rank[i] > 0) {
                int j = sa[rank[i] - 1];
                while (i + h < n && j + h < n &&
                       text.charAt(i + h) == text.charAt(j + h)) {
                    h++;
                }

                lcp[rank[i] - 1] = h;

                if (h > 0) {
                    h--;
                }
            }
        }
    }

    /**
     * Find all occurrences of pattern in text.
     *
     * @param pattern Pattern to find
     * @return List of starting positions (sorted)
     */
    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public List<Integer> findPattern(String pattern) {
        List<Integer> matches = new ArrayList<>();

        if (pattern.isEmpty() || n == 0) {
            return matches;
        }

        // Linear search in suffix array for simplicity
        for (int i = 0; i < n; i++) {
            String suffix = text.substring(sa[i]);
            if (suffix.startsWith(pattern)) {
                matches.add(sa[i]);
            }
        }

        Collections.sort(matches);
        return matches;
    }

    /**
     * Find longest substring that appears at least twice.
     *
     * @return Longest repeated substring
     */
    public String longestRepeatedSubstring() {
        if (n == 0) {
            return "";
        }

        int maxLcp = 0;
        int maxIdx = -1;

        for (int i = 0; i < lcp.length; i++) {
            if (lcp[i] > maxLcp) {
                maxLcp = lcp[i];
                maxIdx = i;
            }
        }

        if (maxIdx == -1) {
            return "";
        }

        int start = sa[maxIdx];
        return text.substring(start, start + maxLcp);
    }

    /**
     * Get the suffix array.
     *
     * @return Copy of suffix array
     */
    public int[] getSuffixArray() {
        return sa.clone();
    }

    /**
     * Get the LCP array.
     *
     * @return Copy of LCP array
     */
    public int[] getLcpArray() {
        return lcp.clone();
    }

    /**
     * [Brief description]
     *
     * @param [param] [description]
     * @return [description]
     * @time O([complexity])
     */
    public static void main(String[] args) {
        // Example 1: Basic suffix array
        System.out.println("=== Example 1: Suffix Array ===");
        String text = "banana";
        SuffixArray sa = new SuffixArray(text);

        System.out.println("Text: \"" + text + "\"");
        System.out.println("Suffix Array: " + Arrays.toString(sa.getSuffixArray()));
        System.out.println("LCP Array: " + Arrays.toString(sa.getLcpArray()));

        System.out.println("\nSuffixes in sorted order:");
        for (int idx : sa.getSuffixArray()) {
            System.out.printf("  %2d: %s%n", idx, text.substring(idx));
        }

        // Example 2: Pattern finding
        System.out.println("\n=== Example 2: Pattern Finding ===");
        String text2 = "mississippi";
        SuffixArray sa2 = new SuffixArray(text2);

        String[] patterns = {"is", "si", "pp", "mis"};
        System.out.println("Text: \"" + text2 + "\"");
        for (String pattern : patterns) {
            List<Integer> positions = sa2.findPattern(pattern);
            System.out.println("Pattern \"" + pattern + "\": positions " + positions);
        }

        // Example 3: Longest repeated substring
        System.out.println("\n=== Example 3: Longest Repeated Substring ===");
        String text3 = "abracadabra";
        SuffixArray sa3 = new SuffixArray(text3);

        String lrs = sa3.longestRepeatedSubstring();
        System.out.println("Text: \"" + text3 + "\"");
        System.out.println("Longest repeated substring: \"" + lrs + "\"");

        String text4 = "aabaabaab";
        SuffixArray sa4 = new SuffixArray(text4);
        String lrs2 = sa4.longestRepeatedSubstring();
        System.out.println("\nText: \"" + text4 + "\"");
        System.out.println("Longest repeated substring: \"" + lrs2 + "\"");
    }
}
